File: String/LongestRepeatingCharacterReplacement/LongestRepeatingCharacterReplacement.py
# 424. Longest Repeating Character Replacement
# https://leetcode.com/problems/longest-repeating-character-replacement/

from collections import defaultdict
from email.policy import default
from typing import Dict


# O(n) two-pointer
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        maxLength = 0
        cMap = defaultdict(int)

        left = 0
        maxCount = 0
        for right, c in enumerate(s):
            cMap[c] += 1
            maxCount = max(maxCount, cMap[c])

            while right - left + 1 - maxCount > k:
                cMap[s[left]] -= 1
                left += 1

            maxLength = max(maxLength, right - left + 1)

        return maxLength


# O(n) two-pointer 매번 최댓값 찾고 검사. 느림.
class Solution1:
    def characterReplacement(self, s: str, k: int) -> int:
        maxLength = 0
        cMap = {}

        left = 0
        length = 0
        for c in s:
            length += 1
            if c in cMap:
                cMap[c] += 1
            else:
                cMap[c] = 1

            # check
            maxChar = max(cMap, key=cMap.get)

            while length - cMap[maxChar] > k:
                if cMap[s[left]] == 1:
                    cMap.pop(s[left])
                else:
                    cMap[s[left]] -= 1
                length -= 1
                left += 1
                maxChar = max(cMap, key=cMap.get)
                print(cMap)

            print(maxChar)

            maxLength = max(maxLength, length)

        return maxLength


def testSol(func, input, output):
    res = func(*input)
    if res == output:
        print(f"O : {res}")
    else:
        print(f"X : {res}	excpet : {output}")


if __name__ == "__main__":
    sol = Solution()

    testSol(sol.characterReplacement, ("ABAB", 2), 4)
    testSol(sol.characterReplacement, ("AABABBA", 1), 4)
    testSol(sol.characterReplacement, ("ABBBABAAA", 2), 6)
    testSol(sol.characterReplacement, ("ABBBABAAA", 2), 5)

File: String/LongestRepeatingCharacterReplacement/LongestRepeatingCharacterReplacement.java
package String.LongestRepeatingCharacterReplacement;
// 424. Longest Repeating Character Replacement
// https://leetcode.com/problems/longest-repeating-character-replacement/

import java.util.Arrays;
import java.util.List;

// O(n)
class Solution {
    public int characterReplacement(String s, int k) {
        int maxLength = 0;
        int[] countChar = new int[26];

        int left = -1;
        int right = 0;
        int maxCount = 0;
        while (right < s.length()) {
            countChar[s.charAt(right) - 'A']++;
            // 현재 윈도우의 최빈값을 찾아야 함.
            // maxCount = Arrays.stream(countChar).max().orElse(0);
            // 하지만 최댓값이 갱신 되지 않으면 maxLength는 갱신 되지 않을 것이므로 굳이 검사할 필요가 없음.
            maxCount = Math.max(maxCount, countChar[s.charAt(right) - 'A']);
            while (right - left - maxCount > k) {
                left++;
                countChar[s.charAt(left) - 'A']--;
            }

            maxLength = Math.max(maxLength, right - left);
            right++;
        }

        return maxLength;
    }
}


public class LongestRepeatingCharacterReplacement {
    public static void main(String[] args) {
        testSol(List.of("ABAB", 2), 4);
        testSol(List.of("AABABBA", 1), 4);
        testSol(List.of("ABBBABAAA", 2), 6);
    }

    static void testSol(List input, Object output) {
        // todo : input, output match
        String arg1 = (String) input.get(0);
        int arg2 = (int) input.get(1);
        int out = (int) output;
        Solution sol = new Solution();
        // todo : solution match
        int res = sol.characterReplacement(arg1, arg2);
        if (res == out) {
            System.out.println("O : " + res);
        } else {
            System.out.println("x : " + res + "	expect : " + out);
        }
    }

}
